### PR TITLE
Add SuperRare trades

### PR DESCRIPTION
**Pull request by catherine-2**
Hi! This is my first non-static table PR so please let me know if there are things I need to fix / any questions! I looked at other trade table files to try to match the format so we can add SuperRare trades.

Brief comments on the purpose of your changes:

*For Dune Engine V2*
I've checked that:
General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributors)
* [ ] if adding a new materialized table, I edited the optimize table macro

Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by soispoke** [comment]
Hi @catherine-2, thanks a lot for you submission :) 

So for NFT platforms, we actually have model we want to follow so we can integrate all the relevant info into the meta table nft.trades/events/mints/burns...
It would be awesome if you could take a look at other similar spells like LooksRare (https://github.com/duneanalytics/spellbook/tree/main/models/looksrare/ethereum) or others if possible ?

**Comment by hildobby** [comment]
Oh I had this one almost done except royalties (was going to work on those before submitting). I planned on finishing it after Zora and Rarible abstractions are merged, let me know if you want to finish it or you want me to take over @catherine-2 :)

**Comment by catherine-2** [comment]
Thank you @soispoke ! Got it - I followed the NFT trades format but will go through and update to make sure it works with all of the different NFT tables so we can add across each. 

@hildobby I will update and tag you to take a look before we merge if that works! I would prefer to work off of this -- spent a lot of time verifying every trade is showing up in this query across all of the different contracts + added logic for RARE trades.

**Comment by catherine-2** [comment]
@soispoke  I think this is ready for re-review! Summary of updates:

- I updated the file structure to match LooksRare and other marketplaces
- Built out the events table which populates fees + trades (and flows into the master nft. tables for each of these)
- Added logic for fees + royalties which weren't previously included
- Updated to use sources instead of the dune.com table names 

I verified that this code is working correctly in a V2 query on dune.com Please let me know if anything looks off. Thank you!

File: models/nft/nft_events.sql [hunk 1]
```diff
@@ -369,6 +369,46 @@ FROM
...
+                platform_fee_amount_usd,
+                platform_fee_percentage,
+                royalty_fee_receive_address,
+                royalty_fee_currency_symbol,
+                royalty_fee_amount_raw,
+                royalty_fee_amount,
+                royalty_fee_amount_usd,
+                royalty_fee_percentage,
+                unique_trade_id
+        FROM {{ ref('superrare_events') }}
```
Comment (soispoke) [inline models/nft/nft_events.sql:410]: replace by `superrare_ethereum_events`

File: models/nft/nft_fees.sql [hunk 2]
```diff
@@ -254,6 +254,41 @@ FROM
...
+                nft_contract_address,
+                project_contract_address,
+                aggregator_name,
+                aggregator_address,
+                tx_hash,
+                block_number,
+                tx_from,
+                tx_to,
+                unique_trade_id
+        FROM {{ ref('superrare_fees') }}
```
Comment (soispoke) [inline models/nft/nft_fees.sql:290]: replace by `superrare_ethereum_fees`

File: models/nft/nft_trades.sql [hunk 3]
```diff
@@ -279,6 +279,36 @@ FROM
...
+                nft_contract_address,
+                project_contract_address,
+                aggregator_name,
+                aggregator_address,
+                tx_hash,
+                block_number,
+                tx_from,
+                tx_to,
+                unique_trade_id
+        FROM {{ ref('superrare_trades') }}
```
Comment (soispoke) [inline models/nft/nft_trades.sql:310]: replace by `superrare_ethereum_trades`

File: models/superrare/superrare_ethereum_events.sql [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+        select  block_time
+                , token_bought_symbol
+                , token_sold_symbol
+                , token_bought_amount
+                , token_sold_amount 
+                , case when token_bought_symbol like '%ETH%' then token_bought_amount*1.0/nullif(token_sold_amount,0) 
+                    else  token_sold_amount*1.0/nullif(token_bought_amount,0)
+                end as eth_per_rare
+        -- from dex.trades
+        from {{ source('dex','trades') }}
```
Comment (soispoke) [inline models/superrare/superrare_ethereum_events.sql:150]: replace by `{{ ref('dex_trades') }}`

File: models/superrare/superrare_ethereum_events.sql [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+            , a.`to` as mint_sent_to
+            , b.`from` as mint_created_by
+    from 
+    (
+    select contract_address
+            , `tokenId`
+            , evt_tx_hash
+            , `to` 
+    -- from erc721_ethereum.evt_Transfer
+    from {{ source('erc721_ethereum','evt_Transfer') }}
```
Comment (soispoke) [inline models/superrare/superrare_ethereum_events.sql:186]: replace by `{{ source('erc721_ethereum','evt_transfer') }}`

File: models/superrare/superrare_ethereum_events.sql [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+            , `from`
+            , lag(`from`) OVER (PARTITION BY contract_address, `tokenId` ORDER BY evt_block_time asc) as previous_owner
+            , case  when `from` = lower('0x8c9f364bf7a56ed058fc63ef81c6cf09c833e656')
+                        then 'From Auction House'
+                    when `to` = lower('0x8c9f364bf7a56ed058fc63ef81c6cf09c833e656')
+                        then 'To Auction House'
+                else '' end as auction_house_flag
+            , ROW_NUMBER() OVER (PARTITION BY contract_address, `tokenId` ORDER BY evt_block_time DESC) AS transaction_rank
+    -- from erc721_ethereum.evt_Transfer
+    from {{ source('erc721_ethereum','evt_Transfer') }}
```
Comment (soispoke) [inline models/superrare/superrare_ethereum_events.sql:229]: replace by `{{ source('erc721_ethereum','evt_transfer') }}`

File: models/superrare/superrare_ethereum_events.sql [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+    from {{ source('ethereum','transactions') }}
+    ) b 
+    on a.evt_tx_hash = b.hash
+)
+, mint_address_per_token_20 as (
+    select contract_address
+            , `value` as `tokenId`
+            , `to` as mint_address 
+    -- from erc20_ethereum.evt_Transfer
+    from {{ source('erc20_ethereum','evt_Transfer') }}
```
Comment (soispoke) [inline models/superrare/superrare_ethereum_events.sql:204]: 'erc20_ethereum','evt_transfer'

File: models/superrare/superrare_ethereum_schema.yml [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+        description:  "Royalty fee in % of the amount paid for a given trade"
+      - &royalty_fee_receive_address
+        name: royalty_fee_receive_address
+        description:  "Wallet addresses receiving fees from the transaction"
+      - &royalty_fee_currency_symbol
+        name: royalty_fee_currency_symbol
+        description: "Symbol of the token in which fees are paid out"
+      - &unique_trade_id
+        name: unique_trade_id
+        description:  "Unique trade ID"
```
Comment (soispoke) [inline models/superrare/superrare_ethereum_schema.yml:127]: add a:        
`tests:
    - unique`

to this column

**Review by soispoke — CHANGES_REQUESTED** [review]
Hi @catherine-2 ! There are still a few small issues, I made comments on the specific files/models but one thing that's also important to add now is incremental filters on the `superrare_ethereum_events` model 🙏

Also, please check the output of the github actions (dbt slim ci) to make sure everything looks good :)

**Comment by catherine-2** [comment]
@soispoke  Thank you for going through and flagging each of the source / ref names that needed to be updated! Sorry about that. 

Just pushed a change to update all of those + add incremental filters. Let me know if there is anything else I should update. Thank you!

**Review by soispoke — COMMENTED** [review]
No problem, thanks @catherine-2 !

You also need to add your project in the dbt_project.yml file (as well as fix the issue I commented about) :)

**Comment by catherine-2** [comment]
@soispoke  I think those are all resolved now! Sorry for all of the minor edits needed on this -- my first non-static table spell :) will make sure to follow all of these practices in future PRs!

**Comment by soispoke** [comment]
Hi @catherine-2, no problem at all very happy that you're learning and contributing to spellbook !

I wanted to push a couple of small changes that still need to be implemented but it seems like I can't do that because I don"t have access to your branch.
Do you mind taking a look at whether you can either accept the changes (I did commit them) or give me permissions ? Otherwise I'll just write it all in another round of review

**Comment by catherine-2** [comment]
@soispoke Where can I see your commits? I already had the box checked here to allow edits to my branch from others but don't see your changes. Let me know if there is another permission I need to update!

File: models/superrare/superrare_ethereum_events.sql [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+                , token_bought_symbol
+                , token_sold_symbol
+                , token_bought_amount
+                , token_sold_amount 
+                , case when token_bought_symbol like '%ETH%' then token_bought_amount*1.0/nullif(token_sold_amount,0) 
+                    else  token_sold_amount*1.0/nullif(token_bought_amount,0)
+                end as eth_per_rare
+        -- from dex.trades
+        from {{ ref('dex_trades') }}
+        where   -- RARE trades
```
Comment (soispoke) [inline models/superrare/superrare_ethereum_events.sql:179]: `{% if is_incremental() %} -- this filter will only be applied on an incremental run
        where block_time >= date_trunc("day", now() - interval '1 week')
        {% endif %}
        and   -- RARE trades`

File: models/superrare/superrare_ethereum_events.sql [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+            , b.`from` as mint_created_by
+    from 
+    (
+    select contract_address
+            , `tokenId`
+            , evt_tx_hash
+            , `to` 
+    -- from erc721_ethereum.evt_Transfer
+    from {{ source('erc721_ethereum','evt_transfer') }}
+    where contract_address in (select distinct contract_address from all_superrare_sales)
```
Comment (soispoke) [inline models/superrare/superrare_ethereum_events.sql:215]: {% if is_incremental() %} -- this filter will only be applied on an incremental run
    and evt_block_time >= date_trunc("day", now() - interval '1 week')
    {% endif %}

File: models/superrare/superrare_ethereum_events.sql [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+    and `tokenId` in (select distinct `_tokenId` from all_superrare_sales)
+    and `from` = '0x0000000000000000000000000000000000000000'
+    ) a 
+    inner join 
+    ( select `from` 
+            , hash 
+    -- from ethereum.transactions
+    from {{ source('ethereum','transactions') }}
+    ) b
```
Comment (soispoke) [inline models/superrare/superrare_ethereum_events.sql:224]: {% if is_incremental() %} -- this filter will only be applied on an incremental run
    where block_time >= date_trunc("day", now() - interval '1 week')
    {% endif %}

File: models/superrare/superrare_ethereum_events.sql [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+    on a.evt_tx_hash = b.hash
+)
+, mint_address_per_token_20 as (
+    select contract_address
+            , `value` as `tokenId`
+            , `to` as mint_address 
+    -- from erc20_ethereum.evt_Transfer
+    from {{ source('erc20_ethereum','evt_transfer') }}
+    where contract_address in (select distinct contract_address from all_superrare_sales)
+    and `value` in (select distinct `_tokenId` from all_superrare_sales)
```
Comment (soispoke) [inline models/superrare/superrare_ethereum_events.sql:234]: {% if is_incremental() %} -- this filter will only be applied on an incremental run
    and evt_block_time >= date_trunc("day", now() - interval '1 week')
    {% endif %}

File: models/superrare/superrare_ethereum_events.sql [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+                        then 'From Auction House'
+                    when `to` = lower('0x8c9f364bf7a56ed058fc63ef81c6cf09c833e656')
+                        then 'To Auction House'
+                else '' end as auction_house_flag
+            , ROW_NUMBER() OVER (PARTITION BY contract_address, `tokenId` ORDER BY evt_block_time DESC) AS transaction_rank
+    -- from erc721_ethereum.evt_Transfer
+    from {{ source('erc721_ethereum','evt_transfer') }}
+    where contract_address in (select distinct contract_address from token_sold_from_auction)
+    and `tokenId` in (select distinct `_tokenId` from token_sold_from_auction)
+    order by 1, 2, 8
```
Comment (soispoke) [inline models/superrare/superrare_ethereum_events.sql:260]: {% if is_incremental() %} -- this filter will only be applied on an incremental run
    and evt_block_time >= date_trunc("day", now() - interval '1 week')
    {% endif %}

File: models/superrare/superrare_ethereum_events.sql [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+            else 'ETH' -- only RARE and ETH possible 
+            end as royalty_fee_currency_symbol,
+        'superrare' || '-' || a.evt_tx_hash || '-' ||  a.`_tokenId`::string || '-' ||  `_seller`::string || '-' || COALESCE(a.contract_address) || '-' || 'Trade'  as unique_trade_id
+
+from all_superrare_sales a
+
+left outer join (select minute, price 
+                -- from prices.usd 
+                from {{ source('prices','usd') }}
+                where blockchain = 'ethereum' and symbol = 'WETH' ) ep
```
Comment (soispoke) [inline models/superrare/superrare_ethereum_events.sql:350]: where blockchain = 'ethereum' and symbol = 'WETH'
                {% if is_incremental() %} -- this filter will only be applied on an incremental run
                and minute >= date_trunc("day", now() - interval '1 week')
                {% endif %}) ep

File: models/superrare/superrare_ethereum_events.sql [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+                where blockchain = 'ethereum' and symbol = 'WETH' ) ep 
+on date_trunc('minute', a.evt_block_time) = ep.minute
+
+left outer join rare_token_price_eth rp
+on date_trunc('week', a.evt_block_time) = rp.week
+
+-- left outer join ethereum.transactions t 
+left outer join  {{ source('ethereum','transactions') }} t
+on a.evt_tx_hash = t.hash
+
```
Comment (soispoke) [inline models/superrare/superrare_ethereum_events.sql:359]: {% if is_incremental() %} -- this filter will only be applied on an incremental run
and t.block_time >= date_trunc("day", now() - interval '1 week')
{% endif %}

File: dbt_project.yml [hunk 16]
```diff
@@ -114,6 +114,12 @@ models:
         +schema: element_avalanche_c
         +materialized: view
 
+    superrare:
+      +schema: superrare
+      +materialized: view
+      ethereum:
```
Comment (soispoke) [inline dbt_project.yml:120]: delete the ethereum part here (nothing under superrare/ethereum path

**Review by soispoke — COMMENTED** [review]
ok I just went ahead and added the changes needed afterall :)

**Comment by soispoke** [comment]
@catherine-2 The dbt slim ci is still failing because of some syntax errors that were introduced by recent commits, Idk if I have been clear enough but by clicking on "details" you can see what errors you need to fix
Capture d’écran 2022-10-06 à 17 35 38

**Comment by soispoke** [comment]
Looking at previous messages I should've said that earlier 😅 does it make sense to you ?

**Comment by catherine-2** [comment]
Yep it's showing this error -- I'm still trying to figure out what it is referring to 😅  will fix 

16:34:10  Runtime Error in model superrare_ethereum_events (models/superrare/superrare_ethereum_events.sql)
16:34:10    [PARSE_SYNTAX_ERROR] Syntax error at or near '='(line 164, pos 38)

**Comment by catherine-2** [comment]
@soispoke  success! fixed the error

**Comment by soispoke** [comment]
@catherine-2 awesome, thanks ! 🎉 everything lgtm now :)

**Review by soispoke — APPROVED** [review]

**Comment by catherine-2** [comment]
🎉 ❤️  thank you @soispoke !

@hildobby FYI this was just merged -- let me know if anything looks off on your end!

**Pull request merged**